### PR TITLE
METRON-831: Add lambda expressions and rudimentary functional programming primitives to Stellar

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
To enable use-cases where the user may want to do simple iteration, we can expose simple functional primitives such as:
* `MAP`
* `REDUCE`
* `FILTER`
To fully support this, we will need lambda expressions as arguments to these functions.

I also added a simple `LIST_ADD` to enable better testing of the functions and because it was an easy-to-fill gap.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron (Incubating).  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root incubating-metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  bin/generate-md.sh
  mvn site:site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommened that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by cestella** [comment]
So, the syntax chosen here for lambda expressions is `&(var_0, var_1, ..., var_n : expr)`.  This seemed like the simplest option to me, but it's possible that we should mimic existing systems such as:
* Pythonish: `lambda var_0, var1, ..., var_n : expr`
* Javaish: `(var_0, var_1, ..., var_n) ->  expr `
* Scalaish: `(var_0, var_1, .., var_n) => expr`

Is there any feedback on the syntax?  Would you like to see something different?

**Comment by nickwallen** [comment]
Really great contribution.  This is going to be so useful.

I agree that it might be better to mimic the syntax from an existing language.  Personally, I think any of the 3 options defined would work well, but based on our roots in Java and the ubiquity of Java, I'd stick with Java's approach.

**Comment by nickwallen** [comment]
Do the higher order functions only work on lists right now?  For example, can I reduce a map?  If not, we can tackle later, but just want to understand the scope.

**Comment by cestella** [comment]
The higher order functions so only work on lists at the moment.  I figured we could add that at contribution a bit later.

**Comment by JonZeolla** [comment]
+1 to sticking with Java(ish) syntax, because I can recall that we've also modeled against Java in other stellar functions.

**Comment by cestella** [comment]
Well, if @JonZeolla then I can't argue. ;)

**Comment by cestella** [comment]
Ok, I migrated to a more java-ish syntax.  I like it better, it's more succinct too, I think.  The following are examples:
* `MAP( [ 1, 2, 3 ], (x) -> x*2)` would yield `[2, 4, 6]`
* `MAP( [ 1, 2, 3], x -> x*2)` would yield `[2, 4, 6]`
* `REDUCE( [1, 2, 3], (sum, x) -> sum + x)` would yield `6`
* `FILTER( [1, 2, 3], () -> true)` would yield `[1, 2, 3]`

**Comment by ottobackwards** [comment]
+1 by review.  The 'assumes one argument' part, actually the 'assumes' part isn't my favorite thing, but we can look at that when we have interactive stellar writing in the ui ( and can validate before deploy ) better sorted out.

**Comment by cestella** [comment]
@ottobackwards Thanks for the +1 and yeah, I definitely hear the concern.  I wanted a rudimentary set of functional primitives, but there should be a follow-up to correct some issues with them around capability, such as supporting more complex data structures than lists (i.e. supporting `MAP({ 'key' : 'value', 'key2' : 'value2' }, (k,v) -> JOIN([k, v], ':'))` should be `['key:value', 'key2:value2'])`).  In that case, it wouldn't assume a single argument.

Oh, also, we do actually validate before a zookeeper push all stellar statements. It will fail if you botch the stellar.  I'm looking forward to a proper UI around stellar, though, for sure. :)

File: metron-platform/metron-common/README.md [hunk 1]
```diff
@@ -466,6 +512,14 @@ The `!=` operator is the negation of the above.
     * IANA Number
   * Returns: The protocol name associated with the IANA number.
 
+### `REDUCE`
+  * Description: Reduces a list by a binary lambda expression. That is, the expression takes two arguments.  Usage example: `REDUCE( [ 1, 2, 3 ] , (x, y) -> x + y)` would sum the input list, yielding `6`.
+  * Input:                      
+    * list - List of arguments.
+    * binary_operation - The lambda expression function to apply to reduce the list. It is assumed that this takes two arguments, the first being the running total and the second being an item from the list.
+    * initial_value - The initial value to use.
```
Comment (jjmeyer0) [inline metron-platform/metron-common/README.md]: This example is missing the `initial_value`.

File: metron-platform/metron-common/README.md [hunk 2]
```diff
@@ -73,6 +73,27 @@ Below is how the `==` operator is expected to work:
...
+
+where 
+* `named_variables` is a comma separated list of variables to use in the Stellar expression
+* `stellar_expression` is an arbitrary stellar expression
+
+
+In the core language functions, we support basic functional programming primitives such as
+* `MAP` - Applies a lambda expression over a list of input.  For instance `MAP([ 'foo', 'bar'], (x) -> TO_UPPER(x) )` returns `[ 'FOO', 'BAR' ]`
+* `FILTER` - Filters a list by a predicate in the form of a lambda expression.  For instance `FILTER([ 'foo', 'bar'], (x ) -> x == 'foo' )` returns `[ 'foo' ]`
+* `REDUCE` - Applies a function over a list of input.  For instance `MAP([ 'foo', 'bar'], (x) -> TO_UPPER(x) )` returns `[ 'FOO', 'BAR' ]`
```
Comment (cestella) [inline metron-platform/metron-common/README.md]: Yep, agreed

File: metron-platform/metron-common/README.md (hunk 1 above)
Comment (cestella) [inline metron-platform/metron-common/README.md]: Yep, agreed

File: metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/FunctionalFunctions.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    @Override
+    public Object apply(List<Object> args) {
+      List<Object> input = (List<Object>) args.get(0);
+      LambdaExpression expression = (LambdaExpression)args.get(1);
+      if(input == null || expression == null) {
+        return input;
+      }
+      List<Object> ret = new ArrayList<>();
+      for(Object o : input) {
+        if(o != null) {
```
Comment (cestella) [inline metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/FunctionalFunctions.java]: You're totally right, I was deferring to `ImmutableList`'s assumptions, which may not be (probably not, in fact) valid here.  I'll adjust accordingly and support nulls.

File: metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/FunctionalFunctions.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    @Override
+    public Object apply(List<Object> args) {
+      List<Object> input = (List<Object>) args.get(0);
+      LambdaExpression expression = (LambdaExpression) args.get(1);
+      if(input == null || expression == null) {
+        return input;
+      }
+      List<Object> ret = new ArrayList<>();
+      for(Object o : input) {
+        if(o == null) {
```
Comment (cestella) [inline metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/FunctionalFunctions.java]: yes, your example should work and return `[null]`

File: metron-platform/metron-common/src/main/java/org/apache/metron/common/stellar/LambdaExpression.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.metron.common.stellar;
+
+import org.apache.metron.common.dsl.Token;
+import org.apache.metron.common.dsl.VariableResolver;
+
+import java.util.*;
```
Comment (cestella) [inline metron-platform/metron-common/src/main/java/org/apache/metron/common/stellar/LambdaExpression.java]: agreed

File: metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/FunctionalFunctions.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+      }
+      LambdaExpression expression = (LambdaExpression) args.get(1);
+
+      Object runningResult = args.get(2);
+      if(expression == null || runningResult == null) {
+        return null;
+      }
+      for(int i = 0;i < input.size();++i) {
+        Object rhs = input.get(i);
+        if(rhs == null) {
```
Comment (cestella) [inline metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/FunctionalFunctions.java]: yep, we should support nulls

**Comment by cestella** [comment]
@jjmeyer0 Thanks so much for the review.  I fixed the support for nulls and added unit tests to test it.  I also adjusted the readme appropriately.

**Comment by jjmeyer0** [comment]
+1 by review. I also spun it up in full dev and tested some basic stellar statements using the new functions. Everything looks good. I think this is a great PR and will be very useful to people.

**Pull request closed without merging**